bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,495 @@
+
+            logger.debug("Connected to mylink at {}", config.ipAddress);
+
+            updateStatus(ThingStatus.ONLINE);
+        } catch (SomfyMyLinkException e) {
+            logger.debug("Problem connecting to mylink, bridge OFFLINE");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    private void ensureKeepAlive() {
+        synchronized (heartbeat) {
+            if (heartbeat == null) {
+                logger.debug("Starting keepalive job in {} min, every {} min", HEARTBEAT_MINUTES, HEARTBEAT_MINUTES);
+                heartbeat = this.scheduler.scheduleWithFixedDelay(this::sendKeepAlive, 1, 1, TimeUnit.MINUTES);
+            }
+        }
+    }
+
+    private void disconnect() {
+        logger.debug("Disconnecting from mylink");
+
+        if (heartbeat != null) {
+            logger.debug("Cancelling keepalive job");
+            heartbeat.cancel(true);
+            heartbeat = null;
+        } else {
+            logger.debug("Keepalive was not active");
+        }
+    }
+
+    private void sendKeepAlive() {
+        try {
+            logger.debug("Keep alive triggered");
+
+            if (getThing().getStatus() != ThingStatus.ONLINE) {
+                // try connecting
+                logger.debug("Bridge offline, trying to connect");
+                connect();
+            } else {
+                // send a ping
+                sendPing();
+                logger.debug("Keep alive succeeded");
+            }
+        } catch (SomfyMyLinkException e) {
+            logger.debug("Problem pinging mylink during keepalive");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    public SomfyMyLinkShade[] getShadeList() throws SomfyMyLinkException {
+        String command = buildShadeCommand("mylink.status.info", "*.*");
+        try {
+            SomfyMyLinkShadesResponse response = (SomfyMyLinkShadesResponse) sendCommandWithResponse(command,
+            SomfyMyLinkShadesResponse.class).get();
+            if (response != null) {
+                return response.getResult();
+            } else {
+                return new SomfyMyLinkShade[0];
+            }
+        }
+        catch (Exception e) {
+            logger.debug("Exception while getting shade list. Message: {}", e.getMessage());
+            return new SomfyMyLinkShade[0];
+        }
+    }
+
+    public void sendPing() throws SomfyMyLinkException {
+        String command = buildShadeCommand("mylink.status.ping", "*.*");
+        sendCommandWithResponse(command, SomfyMyLinkPingResponse.class);
+        return;
+    }
+
+    public SomfyMyLinkScene[] getSceneList() throws SomfyMyLinkException {
+        String command = buildShadeCommand("mylink.scene.list", "*.*");
+
+        try {
+            SomfyMyLinkScenesResponse response = (SomfyMyLinkScenesResponse) sendCommandWithResponse(command,
```

Comment:
Now that generics have been added to `sendCommandWithResponse` I don't think the casting is necessary, but if you had another reason to keep it then feel free to ignore this suggestion.
```suggestion
            SomfyMyLinkScenesResponse response = sendCommandWithResponse(command,
```

Reply:
done